pkg/controller/apiservicecabundle/apiservice_cabundle_controller.go

```diff
@@ -0,0 +1,158 @@
+	apiServiceClient apiserviceclient.APIServicesGetter
+
+	// Services that need to be checked
+	queue workqueue.RateLimitingInterface
+
+	apiServiceLister    apiservicelister.APIServiceLister
+	apiServiceHasSynced cache.InformerSynced
+
+	caBundle []byte
+
+	// syncHandler does the work. It's factored out for unit testing
+	syncHandler func(serviceKey string) error
+}
+
+func NewAPIServiceCABundleInjector(apiServiceInformer apiserviceinformer.APIServiceInformer, apiServiceClient apiserviceclient.APIServicesGetter, caBundle []byte) *ServiceServingCertUpdateController {
+	sc := &ServiceServingCertUpdateController{
+		apiServiceClient: apiServiceClient,
+
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		caBundle: caBundle,
+	}
+
+	sc.apiServiceLister = apiServiceInformer.Lister()
+	apiServiceInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    sc.addAPIService,
+			UpdateFunc: sc.updateAPIService,
+		},
+	)
+	sc.apiServiceHasSynced = apiServiceInformer.Informer().GetController().HasSynced
+
+	sc.syncHandler = sc.syncAPIService
+
+	return sc
+}
+
+func (c *ServiceServingCertUpdateController) syncAPIService(key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	apiService, err := c.apiServiceLister.Get(name)
+	if kapierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	if strings.ToLower(apiService.Annotations[InjectCABundleAnnotationName]) != "true" {
+		return nil
+	}
+	if reflect.DeepEqual(apiService.Spec.CABundle, c.caBundle) {
+		return nil
+	}
+
+	apiServiceToUpdate := apiService.DeepCopy()
+	apiServiceToUpdate.Spec.CABundle = c.caBundle
+	_, err = c.apiServiceClient.APIServices().Update(apiServiceToUpdate)
+	return err
+}
+
+// Run begins watching and syncing.
+func (c *ServiceServingCertUpdateController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Wait for the stores to fill
+	if !cache.WaitForCacheSync(stopCh, c.apiServiceHasSynced) {
+		return
+	}
+
+	glog.V(5).Infof("Starting workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.V(1).Infof("Shutting down")
+}
+
+func (c *ServiceServingCertUpdateController) addAPIService(obj interface{}) {
+	apiService := obj.(*apiregistrationapiv1.APIService)
+	if strings.ToLower(apiService.Annotations[InjectCABundleAnnotationName]) != "true" {
+		return
+	}
+
+	glog.V(4).Infof("adding %s", apiService.Name)
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
+	}
+	c.queue.Add(key)
+}
+
+func (c *ServiceServingCertUpdateController) updateAPIService(old, cur interface{}) {
+	obj := cur.(*apiregistrationapiv1.APIService)
+	if strings.ToLower(obj.Annotations[InjectCABundleAnnotationName]) != "true" {
```

Comment:
@deads2k maybe we can extract this into a helper?

```go
func hasInjectCABundleAnnotation(obj runtime.Object) bool {
  m, _ := meta.Accessor(obj)
  return m.GetAnnotations()[InjectCABundleAnnotationName] == true
}
```

Reply:
> @deads2k maybe we can extract this into a helper?

not really a fan of small helpers like that.